Hook-all-loaders

KarimTayie opened the pull request.

LilyFoote commented on src/template.rs:
```diff
@@ -112,12 +112,115 @@ pub mod django_rusty_templates {
+
+            // Try as tuple (loader, args)
+            if let Ok(loader_tuple) = ld.downcast::<PyTuple>() {
+                if let Ok((loader_obj, args)) =
+                    loader_tuple.extract::<(Bound<'_, PyString>, Bound<'_, PyAny>)>()
```
The [Python code we're re-implementing](https://github.com/django/django/blob/42ab99309d347f617d60751c2e8d627fb2963049/django/template/engine.py#L142-L143) does:

```python
        if isinstance(loader, (tuple, list)):
            loader, *args = loader
```

but the Rust code here only handles the case where `loader` has two elements. I think we should write a helper function:

```rust
fn unpack(loader: Bound<'py, PyAny>) -> PyResult<(Bound<'py, PyString>, Vec<Bound<'py, PyAny>>)> {
    let mut items = loader.try_iter()?;
    let loader_path = items.next().downcast()?;
    Ok((loader_path, items.collect()))
}
```

LilyFoote commented on src/template.rs:
```diff
@@ -112,12 +112,115 @@ pub mod django_rusty_templates {
+                {
+                    let loader_path = loader_obj.extract::<String>()?;
+
+                    // Process args based on type
+                    // Check if args is PyList
+                    if let Ok(true) = args.is_instance(&py.get_type::<PyList>()) {
+                        let args_list = args.downcast::<PyList>()?;
+                        let args_vec = args_list
+                            .iter()
+                            .map(|item| item.extract::<String>())
+                            .collect::<Result<Vec<String>, PyErr>>()?;
+
+                        return Self::map_loader(py, &loader_path, args_vec, HashMap::new());
+                    // Check if args is PyTuple
+                    } else if let Ok(true) = args.is_instance(&py.get_type::<PyTuple>()) {
+                        let args_tuple = args.downcast::<PyTuple>()?;
+                        let args_vec = args_tuple
+                            .iter()
+                            .map(|item| item.extract::<String>())
+                            .collect::<Result<Vec<String>, PyErr>>()?;
+
+                        return Self::map_loader(py, &loader_path, args_vec, HashMap::new());
+                    // Check if args is PyDict
+                    } else if let Ok(true) = args.is_instance(&py.get_type::<PyDict>()) {
+                        let args_dict = args.downcast::<PyDict>()?;
+                        let args_hashmap = args_dict
+                            .iter()
+                            .map(|(key, value)| {
+                                Ok((key.extract::<String>()?, value.extract::<String>()?))
+                            })
+                            .collect::<Result<HashMap<String, String>, PyErr>>()?;
+
+                        return Self::map_loader(py, &loader_path, Vec::new(), args_hashmap);
+                    } else {
+                        return Err(PyErr::new::<pyo3::exceptions::PyTypeError, _>(format!(
+                            "Unsupported args type: {:?}",
+                            args
+                        )));
+                    }
+                }
```
With `unpack` we can do something like:

```suggestion
                let (loader, args) = unpack(loader)?;
                return Self::map_loader(py, loader.to_str()?, args);
```

and we can move the handling of the contents of `args` into `map_loader` where we know which loader we're working with.

LilyFoote commented on src/template.rs:
```diff
@@ -112,12 +112,115 @@ pub mod django_rusty_templates {
+        fn map_loader(
+            py: Python<'_>,
+            loader_path: &str,
+            args_vec: Vec<String>,
+            args_hashmap: HashMap<String, String>,
```
With the `unpack` refactor:

```suggestion
            args: Vec<Bound<'_, PyAny>>,
```

LilyFoote commented on tests/test_engine.py:
```diff
@@ -53,5 +52,83 @@ def test_pathlib_dirs():
+    assert template.render(context) == expected
+
+
+def test_cached_loader_priority():
+    loaders = [
+        (
+            "django.template.loaders.cached.Loader",
+            [
+                "django.template.loaders.filesystem.Loader",
+                "django.template.loaders.app_directories.Loader",
+            ],
+        ),
+    ]
```
We also need to be able to handle this:

```python
    loaders = [
        (
            "django.template.loaders.cached.Loader",
            [
                ("django.template.loaders.filesystem.Loader", ["tests", "app"]),
                "django.template.loaders.app_directories.Loader",
            ],
        ),
    ]
```

LilyFoote reviewed (changes requested):
Thanks for making a start with this!

I've left a bunch of feedback on things I think we can improve.

KarimTayie commented:
> Thanks for making a start with this!
> 
> I've left a bunch of feedback on things I think we can improve.

Thank you for the thorough review and detailed feedback! I really appreciate the time and effort you put into it. I’ll start addressing the comments as soon as I have the time.
